File: Commercial/Commercial_ControllerCS/Commercial_ControllerCS/Elevator.cs
using System;
using System.Collections.Generic;
using static System.Console;

namespace Commercial_ControllerCS
{
    class Elevator
    {
        #region STATIC FIELDS
        private static int _originFloor = 1;
        #endregion

        #region STATIC PROPERTIES
        public static int OriginFloor
        {
            get { return _originFloor; }
            set
            {
                if (value != 1)
                    throw new Exception("Invalid value. The RC (origin) floor should always be at floor 1.");
                else
                    _originFloor = value;
            }
        }
        #endregion

        #region FIELDS
        private string _status;
        private string _movement;
        private int _maxWeightKG;
        private int _currentFloor;
        private int _nextFloor;
        #endregion

        #region PROPERTIES
        public int ID { get; private set; }
        public string Status 
        {
            get { return _status; }
            set
            {
                if (value.ToLower() != "online" && value.ToLower() != "offline")
                    throw new Exception("Invalid value for elevator's status. Can only be either 'online' or 'offline'.");
                else
                    _status = value;
            }
        }
        public string Movement
        {
            get { return _movement; }
            set
            {
                if (value.ToLower() != "up" && value.ToLower() != "down" && value.ToLower() != "idle")
                    throw new Exception("Invalid value for elevator's movement. Can only be either 'up', 'down', or 'idle'.");
                else
                    _movement = value;
            }
        }
        public int MaxWeightKG
        {
            get { return _maxWeightKG; }
            private set
            {
                if (value <= 0)
                    throw new Exception("Invalid max weight limit for elevators. Can't be 0kg or lower.");
                else
                    _maxWeightKG = value;
            }
        }
        public int CurrentFloor 
        {
            get { return _currentFloor; }
            set
            {
                if (value > Battery.NumFloors || value < -(Battery.NumBasements))
                    throw new Exception("The current floor value provided for the elevator is invalid.");
                else
                    _currentFloor = value;
            }
        }
        public int NextFloor 
        {
            get { return _nextFloor; }
            set
            {
                if (value > Battery.NumFloors || value < -(Battery.NumBasements))
                    throw new Exception("The current floor value provided for the elevator is invalid.");
                else
                    _nextFloor = value;
            }
        }

        public Column Column { get; private set; }
        public FloorDisplay FloorDisplay { get; private set; }
        public List<Request> RequestsQueue { get; set; }
        public ElevatorDoor Door { get; set; }
        #endregion

        #region CONSTRUCTOR
        public Elevator(int id_, Column column_)
        {
            ID = id_;
            Column = column_;
            Status = "online"; // online|offline
            Movement = "idle"; // up|down|idle
            CurrentFloor = 1;
            FloorDisplay = new FloorDisplay(this);
            RequestsQueue = new List<Request>();
            Door = new ElevatorDoor("closed");
        }
        #endregion

        #region METHODS
        // Change properties of elevator in one line - USE ONLY FOR TESTING
        public void ChangeProperties(int newCurrentFloor, int newNextFloor)
        {
            CurrentFloor = newCurrentFloor;
            NextFloor = newNextFloor;

            if (CurrentFloor > NextFloor)
                Movement = "down";
            else
                Movement = "up";
                
            RequestsQueue.Add(new Request(NextFloor, Movement));
        }
        public void ChangeProperties(int newCurrentFloor)
        {
            CurrentFloor = newCurrentFloor;
            Movement = "idle";
        }

        // Make elevator go to its scheduled next floor
        public void GoToNextFloor()
        {
            if (CurrentFloor != NextFloor)
            {
                if (CurrentFloor > 0)
                    WriteLine($"Elevator {ID} of Column {Column.ID}, currently at floor {CurrentFloor}, is about to go to floor {NextFloor}...");
                else if (NextFloor < 0)
                    WriteLine($"Elevator {ID} of Column {Column.ID}, currently at floor B{Math.Abs(CurrentFloor)}, is about to go to floor B{Math.Abs(NextFloor)}...");
                else if (NextFloor > 0)
                    WriteLine($"Elevator {ID} of Column {Column.ID}, currently at floor B{Math.Abs(CurrentFloor)}, is about to go to floor {NextFloor}...");
                WriteLine("=================================================================");

                FloorDisplay.DisplayFloor();

                // Traverse through the floors
                while (CurrentFloor != NextFloor)
                {
                    // Do not display floors that are not part of the column's range
                    if (Movement.ToLower() == "up")
                    {
                        if (CurrentFloor + 1 < Column.LowestFloor)
                        {
                            WriteLine($"\n... Quickly traversing through the floors not in column {Column.ID}'s usual elevator range ...\n"); ;
                            CurrentFloor = Column.LowestFloor;
                        }
                        else
                            CurrentFloor++;
                    }
                    else
                    {
                        if (CurrentFloor - 1 < Column.LowestFloor)
                        {
                            WriteLine($"\n... Quickly traversing through the floors not in column {Column.ID}'s usual elevator range ...\n");
                            CurrentFloor = OriginFloor;
                        }
                        else
                            CurrentFloor--;
                    }
                    FloorDisplay.DisplayFloor();
                }

                WriteLine("=================================================================");
                if (CurrentFloor > 0)
                    WriteLine($"Elevator {ID} of Column {Column.ID} has reached its requested floor! It is now at floor {CurrentFloor}");
                else
                    WriteLine($"Elevator {ID} of Column {Column.ID} has reached its requested floor! It is now at floor B{Math.Abs(CurrentFloor)}");
            }
        }

        // Make elevator go to origin floor
        public void GoToOrigin()
        {
            NextFloor = OriginFloor;
            WriteLine($"Elevator {ID} of Column {Column.ID} going back to RC (floor {Elevator.OriginFloor})...");
            GoToNextFloor();
        }

        // Set what should be the movement direction of the elevator for its upcoming request
        private void SetMovement()
        {
            int floorDifference = CurrentFloor - RequestsQueue[0].Floor;

            if (floorDifference > 0)
                Movement = "down";
            else if (floorDifference < 0)
                Movement = "up";
            else
                Movement = "idle";
        }

        // Sort requests, for added efficiency
        private void SortRequestsQueue()
        {
            Request request = RequestsQueue[0];

            // Remove any requests which are useless i.e. requests that are already on their desired floor
            foreach (var req in RequestsQueue.ToArray())
            {

                if (req.Floor == CurrentFloor)
                    RequestsQueue.Remove(RequestsQueue.Find(x => x.Floor == CurrentFloor));
            }

            SetMovement();

            if (RequestsQueue.Count > 1)
            {
                if (Movement == "up")
                {
                    // Sort the queue in ascending order
                    RequestsQueue.Sort((x, y) => x.Floor.CompareTo(y.Floor));

                    // Push any request to the end of the queue that would require a direction change
                    foreach (var req in RequestsQueue.ToArray())
                    {
    
                        if (req.Direction != Movement || req.Floor < CurrentFloor)
                        {
                            RequestsQueue.Remove(req);
                            RequestsQueue.Add(req);
                        }
                    }
                }

                else
                {
                    // Reverse the sorted queue (will now be in descending order)
                    RequestsQueue.Sort((x, y) => y.Floor.CompareTo(x.Floor));

                    // Push any request to the end of the queue that would require a direction change
                    foreach (var req in RequestsQueue.ToArray())
                    {

                        if (req.Direction != Movement || req.Floor > CurrentFloor)
                        {
                            RequestsQueue.Remove(req);
                            RequestsQueue.Add(req);
                        }
                    }
                }

            }
        }

        // Complete the elevator requests
        public void DoRequests()
        {
            if (RequestsQueue.Count > 0)
            {
                // Make sure queue is sorted before any request is completed
                SortRequestsQueue();
                Request requestToComplete = RequestsQueue[0];

                // Go to requested floor
                if (Door.Status != "closed")
                    Door.CloseDoor();
                NextFloor = requestToComplete.Floor;
                GoToNextFloor();

                // Remove request after it is complete
                Door.OpenDoor();
                RequestsQueue.Remove(requestToComplete);

                // Automatically close door
                Door.CloseDoor();
            }
            // Automatically go idle temporarily if 0 requests or at the end of request
            Movement = "idle";
        }

        // Check if elevator is at full capacity
        public void CheckWeight(int currentWeightKG)
        {
            // currentWeightKG calculated thanks to weight sensors
            if (currentWeightKG > MaxWeightKG)
            {
                // Display 10 warnings
                for (var i = 0; i < 10; i++)
                {
                    WriteLine($"\nALERT: Maximum weight capacity reached on Elevator {ID} of Column {Column.ID}");
                }

                // Freeze elevator until weight goes back to normal
                Movement = "idle";
                Door.OpenDoor();
            }
        }
        #endregion
    }
}

File: Commercial/Commercial_ControllerGo/Elevator.go
package main

import (
	"fmt"
	"math"
	"sort"
)

// STATIC PROPERTIES (in Golang, it's treated as a global)
var OriginFloor int = 1

type Elevator struct {

	// PROPERTIES
	ID            int
	Status        string // online|offline
	Movement      string // up|down|idle
	MaxWeightKG   int
	CurrentFloor  int
	NextFloor     int
	Column        *Column
	FloorDisplay  FloorDisplay
	RequestsQueue []Request
	Door          ElevatorDoor
}

// CONSTRUCTOR :-> Initialize an elevator with the default values when creating a column's elevator list
func (elevator *Elevator) InitElevator(id int, column *Column) {

	elevator.ID = id
	elevator.Column = column
	elevator.Status = "online"
	elevator.Movement = "idle"
	elevator.CurrentFloor = 1
	elevator.MaxWeightKG = 2000
	elevator.FloorDisplay = FloorDisplay{elevator}
	elevator.RequestsQueue = []Request{}
	elevator.Door = ElevatorDoor{"closed"}
}

// METHODS
// Change properties of an ACTIVE elevator in one line - USE ONLY FOR TESTING
func (elevator *Elevator) ChangePropertiesActive(newCurrentFloor int, newNextFloor int) {

	elevator.CurrentFloor = newCurrentFloor
	elevator.NextFloor = newNextFloor

	if elevator.CurrentFloor > elevator.NextFloor {
		elevator.Movement = "down"

	} else {

		elevator.Movement = "up"
	}

	r := Request{elevator.NextFloor, elevator.Movement}
	elevator.RequestsQueue = append(elevator.RequestsQueue, r)
}

// Change properties of an IDLE elevator in one line - USE ONLY FOR TESTING
func (elevator *Elevator) ChangePropertiesIdle(newCurrentFloor int) {

	elevator.CurrentFloor = newCurrentFloor
	elevator.Movement = "idle"
}

// Make elevator go to its scheduled next floor
func (elevator *Elevator) GoToNextFloor() {

	if elevator.CurrentFloor != elevator.NextFloor {

		if elevator.CurrentFloor > 0 {

			fmt.Printf("\nElevator %d of Column %d, currently at floor %d, is about to go to floor %d...",
				elevator.ID, elevator.Column.ID, elevator.CurrentFloor, elevator.NextFloor)

		} else if elevator.NextFloor < 0 {

			fmt.Printf("\nElevator %d of Column %d, currently at floor B%d, is about to go to floor B%d...",
				elevator.ID, elevator.Column.ID, int64(math.Abs(float64(elevator.CurrentFloor))), int64(math.Abs(float64(elevator.NextFloor))))

		} else if elevator.NextFloor > 0 {

			fmt.Printf("\nElevator %d of Column %d, currently at floor B%d, is about to go to floor %d...",
				elevator.ID, elevator.Column.ID, int64(math.Abs(float64(elevator.CurrentFloor))), elevator.NextFloor)
		}
		fmt.Println("\n=================================================================")
		elevator.FloorDisplay = FloorDisplay{elevator}
		elevator.FloorDisplay.DisplayFloor()

		// Traverse through the floors
		for elevator.CurrentFloor != elevator.NextFloor {

			// Do not display floors that are not part of the column's range
			if elevator.Movement == "up" {

				if elevator.CurrentFloor+1 < elevator.Column.LowestFloor {

					fmt.Printf("\n\n... Quickly traversing through the floors not in column %d's usual elevator range ...\n",
						elevator.Column.ID)
					elevator.CurrentFloor = elevator.Column.LowestFloor
				} else {
					elevator.CurrentFloor += 1
				}

			} else {

				if elevator.CurrentFloor-1 < elevator.Column.LowestFloor {

					fmt.Printf("\n\n... Quickly traversing through the floors not in column %d's usual elevator range ...\n",
						elevator.Column.ID)
					elevator.CurrentFloor = OriginFloor
				} else {
					elevator.CurrentFloor -= 1
				}
			}

			//elevator.FloorDisplay = FloorDisplay{elevator}
			elevator.FloorDisplay.DisplayFloor()
		}
		fmt.Println("\n=================================================================")

		if elevator.CurrentFloor > 0 {

			fmt.Printf("\nElevator %d of Column %d has reached its requested floor! It is now at floor %d...",
				elevator.ID, elevator.Column.ID, elevator.CurrentFloor)
		} else {

			fmt.Printf("\nElevator %d of Column %d has reached its requested floor! It is now at floor B%d...",
				elevator.ID, elevator.Column.ID, int64(math.Abs(float64(elevator.CurrentFloor))))
		}
	}
}

// Make elevator go to origin floor
func (elevator *Elevator) GoToOrigin() {

	elevator.NextFloor = OriginFloor
	fmt.Printf("\nElevator %d of Column %d going back to RC (floor %d)...",
		elevator.ID, elevator.Column.ID, OriginFloor)
	elevator.GoToNextFloor()
}

// Set what should be the movement direction of the elevator for its upcoming request
func (elevator *Elevator) SetMovement() {

	floorDifference := elevator.CurrentFloor - elevator.RequestsQueue[0].Floor

	if floorDifference > 0 {
		elevator.Movement = "down"
	} else if floorDifference < 0 {

		elevator.Movement = "up"
	} else {

		elevator.Movement = "idle"
	}
}

// Send new request to its request queue
func (elevator *Elevator) SendRequest(stopFloor int, btnDirection string) {

	r := Request{stopFloor, btnDirection}
	elevator.RequestsQueue = append(elevator.RequestsQueue, r)
}

// Sort requests, for added efficiency
func (elevator *Elevator) SortRequestsQueue() {

	// Remove any requests which are useless i.e. requests that are already on their desired floor
	for i, req := range elevator.RequestsQueue {

		if req.Floor == elevator.CurrentFloor {

			elevator.RequestsQueue = append(elevator.RequestsQueue[:i-1], elevator.RequestsQueue[i+1:]...)
		}
	}

	// Decide if elevator is going up, down or is staying idle
	elevator.SetMovement()

	// Sort
	if len(elevator.RequestsQueue) > 1 {

		if elevator.Movement == "up" {

			// Sort the queue in ascending order
			sort.SliceStable(elevator.RequestsQueue, func(i int, j int) bool {
				return elevator.RequestsQueue[i].Floor < elevator.RequestsQueue[j].Floor
			})

			//  Push any request to the end of the queue that would require a direction change
			for i, req := range elevator.RequestsQueue {

				if req.Direction != elevator.Movement || req.Floor < elevator.CurrentFloor {

					elevator.RequestsQueue = append(elevator.RequestsQueue[:i-1], elevator.RequestsQueue[i+1:]...)
					elevator.RequestsQueue = append(elevator.RequestsQueue, req)
				}
			}

		} else {

			// Reverse the sorted queue (will now be in descending order)
			sort.SliceStable(elevator.RequestsQueue, func(i int, j int) bool {
				return elevator.RequestsQueue[j].Floor < elevator.RequestsQueue[i].Floor
			})

			//  Push any request to the end of the queue that would require a direction change
			for i, req := range elevator.RequestsQueue {

				if req.Direction != elevator.Movement || req.Floor > elevator.CurrentFloor {

					elevator.RequestsQueue = append(elevator.RequestsQueue[:i-1], elevator.RequestsQueue[i+1:]...)
					elevator.RequestsQueue = append(elevator.RequestsQueue, req)
				}
			}

		}
	}
}

// Complete the elevator requests
func (elevator *Elevator) DoRequests() {

	if len(elevator.RequestsQueue) > 0 {

		// Make sure queue is sorted before any request is completed
		elevator.SortRequestsQueue()
		requestToComplete := elevator.RequestsQueue[0]

		// Go to requested floor
		if elevator.Door.Status != "closed" {
			elevator.Door.CloseDoor()
		}
		elevator.NextFloor = requestToComplete.Floor
		elevator.GoToNextFloor()

		// Remove request after it is complete
		elevator.Door.OpenDoor()
		elevator.RequestsQueue = append(elevator.RequestsQueue[1:])

		// Automatically close door
		elevator.Door.CloseDoor()
	}
	// Automatically go idle temporarily if 0 requests or at the end of request
	elevator.Movement = "idle"
}

// Check if elevator is at full capacity
func (elevator *Elevator) CheckWeight(currentWeightKG int) {

	// currentWeightKG calculated thanks to weight sensors
	if currentWeightKG > elevator.MaxWeightKG {

		// Display 10 warnings
		for i := 0; i < 10; i++ {
			fmt.Printf("\nALERT: Maximum weight capacity reached on Elevator %d of Column %d",
				elevator.ID, elevator.Column.ID)

		}

		// Freeze elevator until weight goes back to normal
		elevator.Movement = "idle"
		elevator.Door.OpenDoor()
	}
}
